Plutarch/Lift.hs

```diff
@@ -16,41 +16,46 @@ import PlutusCore.Constant (readKnownConstant)
 import PlutusCore.Evaluation.Machine.Exception (MachineError)
 import qualified UntypedPlutusCore as UPLC
 
--- FIXME: `h -> p`
-class (PLifted p ~ h, PLC.DefaultUni `PLC.Includes` PLiftedRepr p) => PUnsafeLiftDecl (h :: Type) (p :: PType) | p -> h where
+class (PConstant (PLifted p), PConstanted (PLifted p) ~ p, PLC.DefaultUni `PLC.Includes` PLiftedRepr p) => PUnsafeLiftDecl (p :: PType) where
   type PLiftedRepr p :: Type
   type PLifted p :: Type
-  pliftToRepr :: h -> PLiftedRepr p
-  pliftFromRepr :: PLiftedRepr p -> Maybe h
+  pliftToRepr :: PLifted p -> PLiftedRepr p
+  pliftFromRepr :: PLiftedRepr p -> Maybe (PLifted p)
 
-class PUnsafeLiftDecl (PLifted p) p => PLift (p :: PType)
-instance PUnsafeLiftDecl (PLifted p) p => PLift (p :: PType)
+class (PLift (PConstanted h), PLifted (PConstanted h) ~ h) => PConstant (h :: Type) where
+  type PConstanted h :: PType
+
+type PLift = PUnsafeLiftDecl
 
 newtype DerivePLiftViaCoercible (h :: Type) (p :: PType) (r :: Type) (s :: S) = DerivePLiftViaCoercible (p s)
+newtype DerivePConstant (h :: Type) (p :: PType) = DerivePConstant h
 
-instance (Coercible h r, PLC.DefaultUni `PLC.Includes` r) => PUnsafeLiftDecl h (DerivePLiftViaCoercible h p r) where
+instance (PConstant h, PConstanted h ~ DerivePLiftViaCoercible h p r, Coercible h r, PLC.DefaultUni `PLC.Includes` r) => PUnsafeLiftDecl (DerivePLiftViaCoercible h p r) where
```

Comment:
hmm so the issue here comes from the `PConstanted (PLifted p) ~ p` constraint not working if the type instances are defined in the way we want for the DerivingVia instance...

If the constraint isn't necessary for another instance, maybe they could be moved to `PLift` ?

Reply:
Yep, that's the solution I've thought about too. I don't think the constraint is necessary for `PConstant`, so I'll try that.

Reply:
I fixed this and will push my finalised changes later.